File: src/main/java/br/furb/polygonpaint/GraphicalObject.java
package br.furb.polygonpaint;

import sun.reflect.generics.reflectiveObjects.NotImplementedException;

import javax.media.opengl.GL;
import java.util.ArrayList;
import java.util.List;

public class GraphicalObject {
    private Color color;
    private GraphicalPrimitive graphicalPrimitive;
    private List<Point4D> points;
    private BoundingBox boundingBox;
    private Transformacao4D transformation;
    private boolean ehAlter;

    public GraphicalObject() {
        setGraphicalPrimitive(GraphicalPrimitive.LINE_LOOP);
        setColor(new Color(0, 0, 0));
        setPoints(new ArrayList<>());
        ehAlter = true;
    }

    public Color getColor() {
        return color;
    }

    public void setColor(Color color) {
        this.color = color;
    }

    public GraphicalPrimitive getGraphicalPrimitive() {
        return graphicalPrimitive;
    }

    public void setGraphicalPrimitive(GraphicalPrimitive graphicalPrimitive) {
        this.graphicalPrimitive = graphicalPrimitive;
    }

    public List<Point4D> getPoints() {
        return points;
    }

    private void setPoints(List<Point4D> points) {
        this.points = points;
    }

    public BoundingBox getBoundingBox() {
        return boundingBox;
    }

    private void setBoundingBox(BoundingBox boundingBox) {
        this.boundingBox = boundingBox;
    }

    public Transformacao4D getTransformation() {
        return transformation;
    }

    public void setTransformation(Transformacao4D transformation) {
        this.transformation = transformation;
    }

    public void selectPolygon() {
        throw new NotImplementedException();
    }

    public void removePolygon() {
        throw new NotImplementedException();
    }

    public void draw(GL gl) {
        gl.glColor3f(getColor().getRed(), getColor().getGreen(), getColor().getBlue());
        gl.glLineWidth(3.0f);
        gl.glBegin(GL.GL_LINES);
        for (Point4D pt : points) {
            gl.glVertex4d(pt.obterX(), pt.obterY(), pt.obterZ(), pt.obterW());
        }
        gl.glEnd();
    }

    public boolean getEhAlter() {
        return ehAlter;
    }

    public void addPoint(Point4D point4D) {
        points.add(point4D);
    }
}

File: src/main/java/br/furb/polygonpaint/GraphicalObject.kt
package br.furb.polygonpaint

import sun.reflect.generics.reflectiveObjects.NotImplementedException
import java.awt.event.MouseEvent

import javax.media.opengl.GL
import java.util.ArrayList

class GraphicalObject {
    var color: Color = Color(0f, 0f, 0f)
    var graphicalPrimitive: GraphicalPrimitive = GraphicalPrimitive.LINE_LOOP
    var points: MutableList<Point4D> = ArrayList()
    var boundingBox: BoundingBox? = null
    var transformation: Transformacao4D? = null
    fun ehAlter() = points.size > 0 && _ehAlter
    private var _ehAlter: Boolean

    init {
        _ehAlter = true
    }

    fun selectPolygon() {
        throw NotImplementedException()
    }

    fun removePolygon() {
        throw NotImplementedException()
    }

    fun draw(gl: GL) {
        gl.glColor3f(0f, 0f, 0f)
        gl.glPointSize(3.0f)
        gl.glLineWidth(3.0f)
        gl.glBegin(primitive)
        for (pt in points) {
            gl.glVertex2d(pt.x, pt.y)
        }
        gl.glEnd()
    }

    private val primitive: Int
        get() = when {
            points.size == 2 && points[0] == points[1] -> GL.GL_POINTS
            else -> graphicalPrimitive.id
        }

    fun selectedPoint() = points.last()

    private fun addPoint(point4D: Point4D) {
        points.add(point4D)
    }

    private val MOUSE_RIGHT_BUTON = 1
    private val MOUSE_LEFT_BUTON = 3

    fun mouseClicked(e: MouseEvent) {
        if(e.button == MOUSE_RIGHT_BUTON){
            if(points.size == 0)
                addPoint(Point4D(e.x.toDouble(), e.y.toDouble(), 0.0, 1.0))
            addPoint(Point4D(e.x.toDouble(), e.y.toDouble(), 0.0, 1.0))
        }else if(points.size > 0){
            points.removeAt(points.size -1)
            _ehAlter = false
        }
    }
}

